pkg/pmem-csi-driver/controllerserver-node.go

```diff
@@ -82,12 +82,13 @@ func NewNodeControllerServer(nodeID string, dm pmdmanager.PmemDeviceManager, sm
 			klog.Warningf("Failed to get volumes: %s", err.Error())
 		}
 		cleanupList := []string{}
-		v := &nodeVolume{}
-		err = sm.GetAll(v, func(id string) bool {
+		v := nodeVolume{}
+		err = sm.GetAll(&v, func(id string) bool {
 			// See if the device data stored at StateManager is still valid
 			for _, devInfo := range devices {
 				if devInfo.VolumeId == id {
 					ncs.pmemVolumes[id] = v.Copy()
+					v = nodeVolume{}
```

Comment:
And what if we don't end up in this location? Won't that skip the resetting of `v`?

This whole API for `GetAll` is becoming very awkward. This doesn't look like the right solution.

It's not possible to keep the state manager completely generic. Let's change the `interface{}` that it works with into a custom interface with a `New` method.

Then `GetAll` can use `dataPtr.New()` to create a clean instance, fill that with `json.Decode`, and pass that to the callback.

Reply:
> And what if we don't end up in this location? Won't that skip the resetting of v?

Yes, this is valid case should be fixed.

Reply:
> This whole API for GetAll is becoming very awkward. This doesn't look like the right solution.


I pushed new revision of the patch, where I changed the sematics of GetAll(), now it returns only available ids,  then caller would use a Get() on those ids. This also reduces unnessasary(stale state entries) file-reads in this case.